### PR TITLE
Backend/manual data set support

### PR DESCRIPTION
**Pull request by GD-Dheer**
This PR refactors the structure of File Upload Controller & Service. These changes were also to accommodate new stakeholder requirements.

Summary of Changes: 
- Introduction of 2 new file types: File Extraction & DataUpload

File Extraction: Handling the old upload endpoint, allowing users to upload png, txt, json, csv & pdf. It has been built with a Factory Design Pattern allowing new file types to easily be extended. Currently, we only support JSON, if the endpoint is provided a valid json file, it will only parse it and return it. All supported file types will have a different implementation of a parser.

DataUpload: API to allow users to upload from manual json form

### CODE REVIEW

File: Server/src/app.ts [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
-import { loadStartupProcess } from './loaders/loadStartupProcess'
+import { LoadStartupProcess } from './loaders/loadStartupProcess'
 
 /**
  * This class contains complete startup procedure of the application. These settings are loaded only once and used
  * to initialize the application. The initial connection to the database is also created here.
  */
 
-let startup = new loadStartupProcess();
+let startup = new LoadStartupProcess();
```
Comment (ds161833) [inline Server/src/app.ts:9]: nice

File: Server/src/controllers/dataExtractionController.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  private fileExtension: string
+
+  constructor(filePath: string, fileExtension: string) {
+    this.filePathOfUpload = filePath;
+    this.fileExtension = fileExtension
+  }
+
+  createRequest(request: Request, response: Response) {
+    if (!request.body) {
+      response.status(400).send({
```
Comment (ds161833) [inline Server/src/controllers/dataExtractionController.ts]: somewhere i think we are sending a string(i think in auth controller) somewhere we are sending an object, just for future to keep in mind to standardize it

Comment (ds161833) [reply Server/src/controllers/dataExtractionController.ts]: an object message i meant

File: Server/src/controllers/dataExtractionController.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        message: "Request Body is empty."
+      });
+    }
+    else {
+      //let fileName: any = request.file.originalname as any;
+      let extension = this.fileExtension.split('.').pop();
```
Comment (ds161833) [inline Server/src/controllers/dataExtractionController.ts]: why do you need to extract extension from a variable called fileExtension? Either extract it in constructor, or name the variable "fileName"

File: Server/src/services/dataProcess/JsonFileHandler.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    getFileHandler(filePath: string): AbstractFileHandler {
+        return new JsonFileHandler(filePath)
+    }
+}
+
+export class JsonFileHandler extends AbstractFileHandler {
+
+    async parseFile(): Promise<IJsonDatasetModel> {
+        try {
+            this.parsedFileData = await JSON.parse(fileSystem.readFileSync(this.filePath))
```
Comment (Arrow105) [inline Server/src/services/dataProcess/JsonFileHandler.ts:23]: parsedFileData doesn't seem to be declared at the start of the class. Oversight?

File: Server/src/controllers/dataExtractionController.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+
+  private async callDataProcessService(filePath: string, extension: string, response: Response): Promise<Response> {
+    let command = 'Extract'
+    let dataService = new DataProcessService(extension, command, null, filePath);
+    try {
+      let extractDataResponse: any = await dataService.extractData();
+      return response.status(extractDataResponse.statusCode).json(extractDataResponse.message);
+    } catch (error) {
+      return response.status(error.status).json(error.message);
```
Comment (ds161833) [inline Server/src/controllers/dataExtractionController.ts:38]: here to add up to the previous comment, somewhere you .json({error: error.message}) and here is just message myabe refactor it all in next pr

File: Server/src/controllers/dataUploadController.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import { DataProcessService } from '../services/dataProcess/dataProcessService'
+
+/**
+ * The dataUploadController is responsible for processing providing instructions to the application if a request comes in
+ * to the /dataupload api. This controller will call the Service for appropriate processing of the input. The
+ * controller is resposible for providing a response back to the Client on success. The fileUploadController 
+ * creates a request for the fileUploadService to process, only passing the path of the file.
+ */
+
+export class DataUploadController {
```
Comment (ds161833) [inline Server/src/controllers/dataUploadController.ts:11]: I see al controllers have a createRequest method, consider putting it into an abstract class if such exist in typescript, and have all controllers inherit it

File: Server/src/controllers/dataUploadController.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    async createRequest(request: Request, response: Response): Promise<Response> {
+        if (!request.body) {
+            response.status(400).send({
+                message: "No Json to Upload Received"
+            });
+        }
+        else {
+            let command = 'Upload'
+            let dataType = 'json'
```
Comment (ds161833) [inline Server/src/controllers/dataUploadController.ts]: why is dataType hardcoded here?

File: Server/src/controllers/dataUploadController.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    async createRequest(request: Request, response: Response): Promise<Response> {
+        if (!request.body) {
+            response.status(400).send({
+                message: "No Json to Upload Received"
+            });
+        }
+        else {
+            let command = 'Upload'
```
Comment (ds161833) [inline Server/src/controllers/dataUploadController.ts]: consider exporting command constants from a separate file, that way its easier to understand what kinda "Commands" exist when you see them all in one file

File: Server/src/routes/dataExtractionRouter.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+let upload = multer({ dest: 'upload', limits: { fileSize: 8000000 } });
+let router = Router();
+
+/**
+ * This route will accept a file and forward to the router. It is first processed by multer middleware,
+ * and the file is stored in a temporary directory called tmp/json. This route is referred for processing by 
+ * the service.
+ */
+router.post('/api/v1/dataExtract',/**  [JWTAuthenticator.verifyJWT, **/upload.single('file'), (request: MulterRequest, response: Response, next: NextFunction) => {
```
Comment (ds161833) [inline Server/src/routes/dataExtractionRouter.ts]: a public post? do you think its not a vulnerability?

Comment (ds161833) [inline Server/src/routes/dataExtractionRouter.ts]: also like the api versioning, good job

File: Server/src/routes/dataExtractionRouter.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  try {
+    let dataExtract = new dataExtractionController(request.file.path, request.file.originalname);
```
Comment (ds161833) [inline Server/src/routes/dataExtractionRouter.ts:21]: i dont know how dependency injection is done in JS but we might wanna learn about it when we wanna pump up code quality

File: Server/src/routes/dataUploadRouter.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+import { Request, Response, Router, NextFunction } from 'express';
+import { DataUploadController } from '../controllers/dataUploadController';
+import { JWTAuthenticator } from '../middleware/JWTAuthenticator';
+
+let router = Router();
+
+/**
+ * This route will accept json data and forward to the controller to verify request. 
+ */
+router.post('/api/v1/dataUpload', /**JWTAuthenticator.verifyJWT, **/(request: Request, response: Response) => {
```
Comment (ds161833) [inline Server/src/routes/dataUploadRouter.ts]: again a commented out token verification, why

File: Server/src/services/dataProcess/IDataProcessService.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+export abstract class FileHandlerFactory {
+
+    abstract getFileHandler(filePath?: string): AbstractFileHandler
+}
+
+export abstract class AbstractFileHandler {
```
Comment (ds161833) [inline Server/src/services/dataProcess/IDataProcessService.ts]: file name doesnt correspond to file contents

File: Server/src/services/dataProcess/JsonFileHandler.ts [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+import { Authors } from '../../models/entities/Authors';
+import { IMaterials } from '../../models/interfaces/MaterialsInterface';
+import { IAuthors } from '../../models/interfaces/AuthorsInterface';
+import { BadRequest } from '@tsed/exceptions';
+import { validationSchema } from '../helpers/validationSchema';
+import { IJsonDatasetModel } from '../../genericInterfaces/DataProcessInterfaces'
+const fileSystem = require('fs');
+
+
+export class JsonFileFactory extends FileHandlerFactory {
```
Comment (ds161833) [inline Server/src/services/dataProcess/JsonFileHandler.ts]: JsonFileHandlerFactory?

File: Server/src/services/dataProcess/JsonFileHandler.ts [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+            this.parsedFileData = await JSON.parse(fileSystem.readFileSync(this.filePath))
+            return this.parsedFileData
+        } catch (err) {
+            throw new BadRequest("Cannot parse your file. Something is wrong with it");
+        }
+    }
+
+    async validateExtractedData(jsonData: any): Promise<JSON> {
+        try {
+            await validationSchema.validate(jsonData)
```
Comment (ds161833) [inline Server/src/services/dataProcess/JsonFileHandler.ts]: i thought we agreed on no validation? you still think it should be present?

File: Server/src/services/dataProcess/JsonFileHandler.ts [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+        try {
+            referenceTypeID = await uploadModel.insertReferenceType(referenceType); console.log('Received reference ID' + referenceTypeID);
+        } catch (err) {
+            console.log('rejected request for referenceTypeID');
+        }
+
+        referencePublisher = this.parsedFileData.reference.publisher;
+        try {
+            publisherNameId = await uploadModel.insertPublisher(referencePublisher);
+            console.log('Received publisher name ID' + publisherNameId);
```
Comment (ds161833) [inline Server/src/services/dataProcess/JsonFileHandler.ts]: wanna preserve the logs?

File: Server/src/services/dataProcess/JsonFileHandler.ts [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+        dataSetComments = this.parsedFileData.data.comments;
+        try {
+            datasetID = await uploadModel.insertFullDataSet(dataSetName, dataSetDataTypeID, publicationID, categoryIDs, allMaterials, dataSetComments)
+            console.log('DatasetID received: ' + datasetID);
+        } catch (err) {
+            console.log('error receiving datasetID....request rejected');
+        }
+
+        //run check on variable vs contents length to see if they're equal
+        if (this.parsedFileData.data.variables.length == this.parsedFileData.data.contents[0].point.length) {
```
Comment (ds161833) [inline Server/src/services/dataProcess/JsonFileHandler.ts]: these checks can be handled by Yup

File: Server/src/controllers/dataUploadController.ts (hunk 6 above)
Comment (GD-Dheer) [reply Server/src/controllers/dataUploadController.ts:11]: This is a good idea. Will flag this for future date

File: Server/src/services/dataProcess/dataProcessService.ts [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+    this.command = command
+    this.filePath = filePath
+    this.jsonBody = jsonBody
+  }
+
+  async extractData(): Promise<IResponse> {
+
+    let requestResponse: IResponse = {} as any
+    switch (this.dataType) {
+      case 'json': {
```
Comment (silent-po2) [inline Server/src/services/dataProcess/dataProcessService.ts]: Good preparation for future file type uploads.

File: Server/src/services/dataProcess/JsonFileHandler.ts [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+            publisherNameId = await uploadModel.insertPublisher(referencePublisher);
+            console.log('Received publisher name ID' + publisherNameId);
+        } catch (err) {
+            console.log('rejected request for inserting publisherNameId')
+        }
+
+
+        referenceAuthors = this.parsedFileData.reference.authors;
+        let allAuthors: Authors[];
+        try {
```
Comment (ds161833) [inline Server/src/services/dataProcess/JsonFileHandler.ts]: i would prefer if all the fields that you are uploading would have their own functions, this function is very long, and hard to keep it all in mind

File: Server/src/genericInterfaces/DataProcessInterfaces.ts [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+    comments: string;
+}
+
+export interface IData {
+    variables: IVariable[];
+    contents: IContent[];
+    comments: string;
+}
+
+export interface IJsonDatasetModel {
```
Comment (silent-po2) [inline Server/src/genericInterfaces/DataProcessInterfaces.ts]: A+ matching the various fields to interfaces 1-to-1

File: Server/src/controllers/dataUploadController.ts (hunk 7 above)
Comment (GD-Dheer) [reply Server/src/controllers/dataUploadController.ts]: The datatype for upload will always be the same

File: Server/src/loaders/loadStartupProcess.ts [hunk 20]
```diff
@@ -93,6 +96,11 @@ export class loadStartupProcess {
       try {
         await connectDB(this.config);
       } catch (error) {
+        // If AlreadyHasActiveConnectionError occurs, return already existent connection
+        if (error.name === "AlreadyHasActiveConnectionError") {
+          const existentConn = getConnectionManager().get("default");
```
Comment (silent-po2) [inline Server/src/loaders/loadStartupProcess.ts:101]: Good catch. This isn't something we had before and it's a solid addition.

File: Server/src/services/dataProcess/JsonFileHandler.ts [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+        material = this.parsedFileData.material;
+        let allMaterials: any[];
+        try {
+            allMaterials = await uploadModel.insertMaterial(material);
+            console.log('received material(s)' + material);
+        } catch (err) {
+            console.log('material(s) not found');
+        }
+
+        category = this.parsedFileData.category;
```
Comment (ds161833) [inline Server/src/services/dataProcess/JsonFileHandler.ts]: Im seeing that you are reassigning all the variables, maybe consider also defining them when they are needed rather then in advance?

File: Server/src/controllers/dataUploadController.ts (hunk 8 above)
Comment (GD-Dheer) [reply Server/src/controllers/dataUploadController.ts]: A-OK

File: Server/src/services/dataProcess/IDataProcessService.ts (hunk 12 above)
Comment (silent-po2) [inline Server/src/services/dataProcess/IDataProcessService.ts]: Again, good work preparing for potential future file types

File: Server/src/routes/dataExtractionRouter.ts (hunk 9 above)
Comment (GD-Dheer) [reply Server/src/routes/dataExtractionRouter.ts]: I mentioned this in chat. This was to be uncommmented after approval

File: Server/src/services/dataProcess/JsonFileHandler.ts [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+        subcategory = this.parsedFileData.subcategory;
+        let categoryIDs = await uploadModel.insertCategories(category, subcategory);
+
+        dataType = this.parsedFileData["data type"];
```
Comment (ds161833) [inline Server/src/services/dataProcess/JsonFileHandler.ts]: i think we can talk with stakeholder he wouldnt mind having data_type with underscore

File: Server/src/routes/dataUploadRouter.ts (hunk 11 above)
Comment (GD-Dheer) [reply Server/src/routes/dataUploadRouter.ts]: Above

File: Server/src/services/dataProcess/IDataProcessService.ts (hunk 12 above)
Comment (GD-Dheer) [reply Server/src/services/dataProcess/IDataProcessService.ts]: Will update

File: Server/src/services/dataProcess/JsonFileHandler.ts (hunk 14 above)
Comment (GD-Dheer) [reply Server/src/services/dataProcess/JsonFileHandler.ts]: This is after the post. This code only gets executed on 'upload'

File: Server/src/services/dataProcess/JsonFileHandler.ts [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+
+
+export class JsonFileFactory extends FileHandlerFactory {
+
+    getFileHandler(filePath: string): AbstractFileHandler {
+        return new JsonFileHandler(filePath)
+    }
+}
+
+export class JsonFileHandler extends AbstractFileHandler {
```
Comment (silent-po2) [inline Server/src/services/dataProcess/JsonFileHandler.ts]: A+ use of interfaces

File: Server/src/tests/controllers/dataUploadController.spec.ts [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+            json: jest.fn(),
+            send: jest.fn()
+        }
+
+    })
+    afterEach(async () => {
+        await getConnection().close();
+    });
+
+    test('Valid Json Upload', async () => {
```
Comment (ds161833) [inline Server/src/tests/controllers/dataUploadController.spec.ts:26]: good job with tests

**Review by silent-po2 — COMMENTED** [review]
Generally, this all looks good

**Review by silent-po2 — CHANGES_REQUESTED** [review]
...\Client\src\Components\DataCell\DataCell.js file needs updating for the new API

**Review by silent-po2 — APPROVED** [review]
Thanks for fixing this. That covers the issue I had with this PR

**Comment by silent-po2** [comment]
I tested this code with an undefined subcategory via Postman; it successfully wrote to the database. Picture below:
image

**Review by silent-po2 — CHANGES_REQUESTED** [review]
Of course, while the code works, there's a slight fix that is needed for the tests. From what I can see, the test failing in travis, ...src/tests/controllers/dataUploadController.spec.ts, exists but under a different name now I believe?

File: Server/src/services/dataProcess/JsonFileHandler.ts (hunk 21 above)
Comment (GD-Dheer) [reply Server/src/services/dataProcess/JsonFileHandler.ts]: Done

File: Server/src/services/dataProcess/JsonFileHandler.ts (hunk 18 above)
Comment (GD-Dheer) [reply Server/src/services/dataProcess/JsonFileHandler.ts]: A+

**Comment by GD-Dheer** [comment]
> Of course, while the code works, there's a slight fix that is needed for the tests. From what I can see, the test failing in travis, ...src/tests/controllers/dataUploadController.spec.ts, exists but under a different name now I believe?

Fixed

**Review by silent-po2 — APPROVED** [review]
Thanks for the fix, approved

File: Server/src/services/dataProcess/JsonFileHandler.ts [hunk 25]
```diff
@@ -36,6 +36,115 @@ export class JsonFileHandler extends AbstractFileHandler {
...
+
+        for (let i = 0; i < dataContentArray.length; i++) {
+            dataPointsForVariable.push(dataContentArray[i].point[index]);
+            dataSetComments.push(dataContentArray[i].comments);
+        }
+        let contentsArrayInfo = [dataPointsForVariable, dataSetComments];
+        return contentsArrayInfo;
+    }
+
+    private async insertUnits(uploadModel, units) {
```
Comment (silent-po2) [inline Server/src/services/dataProcess/JsonFileHandler.ts]: Moving these methods here is a good tweak

**Review by ds161833 — APPROVED** [review]
NIce, I really like how we cleaned up the code, also the factory there is looking reeealy good, A+

**Pull request merged**